### PR TITLE
Add trailgen mapgen: v6-like with biome API support

### PR DESCRIPTION
**Pull request by Cartmic**
This pull request is to include a mapgen which is similar to v6 in the general shape the world but uses the biome API instead of v6's hard-coded biomes.

I wasn't sure where to put trailgen in the list of mapgens so I left it at the bottom of the list under v6, but let me know where you'd like it and I can move it.

Trailgen's forum thread is here: https://forum.minetest.net/viewtopic.php?f=7&t=26751

Kind regards,

Cartmic

### CODE REVIEW

**Comment by Cartmic** [comment]
I based this on 5.6.1, it seems merging it with master didn't work, I'm redoing the files now on master.

**Comment by TurkeyMcMac** [comment]
I have noticed places where mesas cut off other terrain:

screenshot_20221203_155505
screenshot_20221203_155359

**Comment by Cartmic** [comment]
> I have noticed places where mesas cut off other terrain:
Thanks for testing it.
What's the seed and location of that?

**Comment by TurkeyMcMac** [comment]
You can find one in seed 433213232 at (-470, -3, -202).

**Comment by Cartmic** [comment]
> 433213232

Thanks.

I checked it out on v6 and it happens there too, so it's proably okay.

433213232v6screenshot

File: src/settings_translation_file.cpp
Comment (sfan5) [inline src/settings_translation_file.cpp:804]: this file is regenerated once in a while, no need to apply any changes now.

File: builtin/settingtypes.txt
Comment (sfan5) [inline builtin/settingtypes.txt:814]: should trailgen be marked as unstable here?

File: src/settings_translation_file.cpp
Comment (Cartmic) [reply src/settings_translation_file.cpp:804]: Ah okay thank you.

File: builtin/settingtypes.txt
Comment (Cartmic) [reply builtin/settingtypes.txt:814]: Yes, I think it would be good to mark it as unstable for now until more people play with it.

**Comment by Cartmic** [comment]
> I have a question. What can you change about trailgen in the settings? Does it have the same customisation options as mapgen v6? Or are there some differences?

The standard basic stuff of a modern mapgen, the cave and dungon values and noise values. No it doesn't have the same customisation values as v6, e.g. the mudflow option, beach and biome values. There is a setting not found in v6 and that is the map generation level setting, which cna generate the world slightly higher or lower depending on your preference e.g. a value of 2 will bring it in line with a previously generated v6 world if you were to drop the map database file into a new trailgen world folder.

**Comment by Cartmic** [comment]
> > The standard basic stuff of a modern mapgen, the cave and dungon values and noise values. No it doesn't have the same customisation values as v6, e.g. the mudflow option, beach and biome values.
> 
> Is that because trailgen uses the modern biome system that all the other mapgens besides mapgen v6 use?

That's correct, the beach and mudflow are tied to specific nodes that would, in a moden game, be biome specific.

> > There is a setting not found in v6 and that is the map generation level setting, which cna generate the world slightly higher or lower depending on your preference e.g. a value of 2 will bring it in line with a previously generated v6 world if you were to drop the map database file into a new trailgen world folder.
> 
> So the default value of this would be 2 in order to make converting v6 worlds to trailgen painless? Or was that just an example?

It was an example but indeed 2 makes the transition smoother.

**Comment by TurkeyMcMac** [comment]
To implement mud flow, I believe you can get the top nodes and filler nodes from the biome information. You could also just use nodes from the `soil` and `sand` groups, perhaps.

Why not set the map generation level to 2 by default, to increase the similarity to v6?

**Comment by Pantyhose-X** [comment]
All mapgen mods show the mapgen option when creating a world

**Comment by Cartmic** [comment]
> To implement mud flow, I believe you can get the top nodes and filler nodes from the biome information. You could also just use nodes from the `soil` and `sand` groups, perhaps.

It's certainly something to look at, I've had some thoughts in the past how I would implement it, I think node groups sound like a good idea, thanks TurkeyMcMac. For now I'd like to leave it out of the current release because it does what it says on the tin; it's a v6-like map-gen rather than a v6-clone.

> Why not set the map generation level to 2 by default, to increase the similarity to v6?

It would be the logical thing to do, however from my tests I've found there is more of a coast line when using values of 2 which is why I set the default as 1 as a nice in between for those who don't like that. I could proably look at implementing a tick box on main world generation screen, which would be a bit more user friendly.

**Comment by Cartmic** [comment]
> > It would be the logical thing to do, however from my tests I've found there is more of a coast line when using values of 2 which is why I set the default as 1 as a nice in between for those who don't like that. I could proably look at implementing a tick box on main world generation screen, which would be a bit more user friendly.
> 
> I would like to see some screenshots of said coastlines. A comparison between trailgen with value 1, trailgen with value 2, and v6 as a reference point. All at the same coordinates and seed.

Here you go:

Seed: 12170282024851156146
You can find it looking east at location: 82,16,1155

v6:
v6screenshot
Trailgen map generation height 0:
trailgen_terrain_height_0
Trailgen map generation height 1:
trailgen_terrain_height_1
Trailgen map generation height 2:
trailgen_terrain_height_2

**Comment by Cartmic** [comment]
> Personally, I still think the default should be 2 instead of 1.

Noted, see my reply below.

>     * Changing the height doesn't really solve the coastline "problem" at all. In fact, all this seems to do is bring the coastline more inland while at the same time flooding everything slightly.

Which I personally like.
 
>     * These coastlines are likely just a result of the mapgen using the modern method of defining biomes instead of having predefined biomes. Modifying the biomes is probably a better way to remedy this "problem".

Indeed it's possible that there's a way of fiddling with the noise values.

>     * Converting worlds from mapgen v6 to trailgen should require taking as few steps as possible. What if someone forgets to change the height to 2 while converting their v6 world?

Noted see below.
 
> The only argument I really saw for changing the height to 1 was that some people "didn't like the coastlines", but I personally think that bigger coastlines look very nice. Not to mention that the coastline in this picture doesn't even look close to being a big coastline, especially when compared to some of the coastlines that mapgen v5 and v7 produce:
> 
> V5 Coastline
> 
> This coastline is about three times larger, and that is ignoring that there's even more coastline to the right (covered in snow) that I didn't show in this picture. I don't really hear people complain about those coastlines.

Yes it's certainly personal opinion.

> If anything, the tickbox should be to set the height to 1 while the default height is 2, instead of having the default height be 1 and have the tickbox set it to 2.

I certainly see your point and thank you for it. The reason why I was hesitant was that I didn't want to see Trailgen simply become a mapgen that is used for transitioning to the biome api from a v6 world. It certainly makes sense to default to level 2 for for the purposes of those people who do wish to bring their world into the new api, however originally dropping a v6 world into Trailgen wasn't in the plan, if I recall correctly, I just happened to be surprised when went back and fiddled with the hardcoded value that it matched with the v6 land so just before I uploaded it to Github I added in that setting. I had considered a tick box for a "v6 compatibility" on the map creation screen, granted that box could be changed to do reverse. I think I'd like to do a vote somewhere, maybe on the forum thread.

**Comment by v-rob** [comment]
Oh, I love v6's terrain!  I always wanted to see a modern v6 mapgen with full biome support.

This mapgen is described as v6-like.  What are the precise differences?  Some differences are clear, like the lack of hardcoded biomes and mudflow and other v6-specific stuff.  Are there any differences in terrain due to different perlin noise parameters or anything?  Does it have the same caves as v6?  Etc.

**Comment by Cartmic** [comment]
> Oh, I love v6's terrain! I always wanted to see a modern v6 mapgen with full biome support.
> 
> This mapgen is described as v6-like. What are the precise differences? Some differences are clear, like the lack of hardcoded biomes and mudflow and other v6-specific stuff. Are there any differences in terrain due to different perlin noise parameters or anything? Does it have the same caves as v6? Etc.

Hi v-Rob! It currently uses the same caves as v7, I could add in v6 caves.
The noise parameters are the same. I have plans for some form of mountains and I occasionally play around with the code with that as my interest takes me. I decided I didn't really want it to simply be a mapgen people drop v6 worlds into, to a certain extent I want it to be it's own thing. I've been using it for my worlds for well over year now I feel like it's ready for merging.

Regarding the v6 terrain level posts above yours, before it had a changable value I was trying to match it up by eye i.e. not dropping merging a v6 and trailgen worlds as much as I could and I settled a 1 and I've been playing my wolds like that since, it was only just before I had opened a merge request that I noticed that a value of 2 actually matches the land up, the beaches don't match up but that's minor a thing as they are seperate as you know. I'd like to keep it as level 1 as the default, as I feel anyone who is going to take the effort of changing a v6 world into trailgen world has capability of ticking a compatbility box(which of course I need to add intead of the current numerical setting).

**Comment by v-rob** [comment]
> It currently uses the same caves as v7, I could add in v6 caves.

Actually, as I now recall, v7 supports v6's caves (they're the settings with "small caves" in the name), but they're disabled by default, as seen in #8928.  The code in this PR seems to have the same settings for trailgen, so it looks like there's no additional work that needs to be done here.

> I decided I didn't really want it to simply be a mapgen people drop v6 worlds into, to a certain extent I want it to be it's own thing.

Not a bad thing; we don't need a perfect v6 compatibility mapgen immediately.  For now, we still have v6 for older worlds, so I think that it's fine to have a little bit of liberty with trailgen, given that we can always add compatibility features later as desired.

> I'd like to keep it as level 1 as the default, as I feel anyone who is going to take the effort of changing a v6 world into trailgen world has capability of ticking a compatbility box.

Given that the caves default to v7 caves by default, not to mention mapgen differences, a level of 1 vs. 2 is probably one of the lesser compatibility problems between the mapgens that will affect worlds.  I wouldn't advise making a compatibility checkbox that might be misleading since people might think that their world will seamlessly convert from one to the other, which it wouldn't due to the serious differences.

I'd have to actually play around in the worlds to be sure, but I doubt that it'll make much difference at all to me whether the level is 1 or 2 by default.

**Comment by v-rob** [comment]
I let this stagnate too long.  Will review within the week.

**Comment by v-rob** [comment]
I started reviewing two days ago, but found that there were a lot formatting errors and C++ issues, plus a bug or two.  Instead of sending a barrage of review items, I just fixed those things in a separate branch: https://github.com/v-rob/minetest/tree/trailgen2.  The changes are in this commit: https://github.com/v-rob/minetest/commit/49ff0e5d6616af1acdeb959dcf12139eb7181e89.

Some notes:

* I moved trailgen up in the mapgen list, right after carpathian and valleys, but before v5 (since I think trailgen/v6 is, in general, a more tame and desirable mapgen than v5).  v6 is at the bottom because of its biome problems, so trailgen belongs further up.
* I do not think that trailgen should be marked as unstable.  It should be a stable replacement for v6, and users shouldn't have to worry about it changing.  If we want to add new features in the future (such as a way to have mudflow), a new flag should just be added (just as caverns have a flag that can be disabled).
* I don't there's any point in having a v6 compatibility checkbox.  If people want to convert v6 maps, they'll have to manually fiddle with their world's settings, changing the mapgen and parameters manually anyway.  New worlds should use the new biomes and caves by default.
* That said, I do believe v6 should be deprecated, and dropped in 6.0.0.  The different biome system is legacy cruft at this point.

Now, my opinion on `mgtrailgen_map_height_mod`.  In the first place, this setting has strictly duplicate functionality: you can modify the `offset` field of the `mgtrailgen_np_terrain_base` noise setting to offset the height of the world.  This is the standard way to do it in the other mapgens, so `mgtrailgen_map_height_mod` should be removed.

Moreover, I think that the terrain should use the exact same parameters and code as v6, i.e. identical coastline height.  Trailgen is, for all intents and purposes, v6 with biomes, so it's more confusing than anything else to the users of the mapgen to have the defaults be slightly different.  Those who like different coastlines can change the terrain offset to suit their personal preference--it's only a few clicks in the settings dialog, after all.

**Comment by Cartmic** [comment]
Thank you for the review v-rob.  Okay I agree with what you said, thanks for putting me straight with regards to the mgtrailgen_np_terrain_base offset field. I'll remove mgtrailgen_map_height_mod.

**Comment by Cartmic** [comment]
I'm fairly in experienced with git, but I'll give it a go and let you know how it goes!

**Comment by Cartmic** [comment]
Okay, so I manually added your fixes v-rob and uploaded them manually too. I had some issues regenerating the settings_translation_file.cpp, I've managed it before but it just doesn't seem to be playing ball, so I edited an exisiting one against a the one in the master branch, I realise it's proably not the best way to go about that but it was last thing preventing me from uploading the changes.

**Comment by grorp** [comment]
Ideally you just remove any changes to `settings_translation_file.cpp`. Changing that file is unnecessary because it is regenerated before release.

**Comment by Cartmic** [comment]
Ah thanks, that's now been replaced with the current one in the master branch.

**Comment by grorp** [comment]
Now you've added a copy of the file to the root folder of the repository.

**Comment by Cartmic** [comment]
Okay that's now removed.

**Comment by ancientmarinerdev** [comment]
This looks really good. How close is it, @v-rob ? Apologies for the ping, but wasn't sure if you'd see this otherwise.

File: builtin/mainmenu/dlg_create_world.lua
Comment (v-rob) [inline builtin/mainmenu/dlg_create_world.lua:76]: The moved code doesn't include all these flags (`temples` was added in a recent commit, I presume)

File: builtin/settingtypes.txt
Comment (v-rob) [inline builtin/settingtypes.txt:1077]: Another rebase mistake; this shouldn't have been changed.

File: builtin/settingtypes.txt
Comment (v-rob) [inline builtin/settingtypes.txt:1104]: This line should not have been changed.  I don't know how this change snuck in, but it could break old v6 worlds (very slightly, but a breakage nonetheless).

Edit: This was my mistake.  That change got into my branch earlier.  Don't know how or why...

File: builtin/settingtypes.txt
Comment (v-rob) [inline builtin/settingtypes.txt:1641]:
```suggestion
mgtrailgen_np_height_select (Height select noise) noise_params_2d 0.5, 1, (250, 250, 250), 4213, 5, 0.69, 2.0, eased
```

See the above.  This line should be updated to match.

**Review by v-rob — APPROVED** [review]
I forgot to approve this, but it does have my approval after the changes made earlier.

Recent rebases have caused a few mistakes to creep in, however.  I've noted these below.  These must be fixed before this can be merged, but otherwise everything looks fine.

File: builtin/settingtypes.txt
Comment (rudzik8) [reply builtin/settingtypes.txt:1104]: Aren't these added to `map_meta.txt` either way? I don't see how this could break older worlds (even though this change is obv unnecessary for this PR).

Comment (v-rob) [reply builtin/settingtypes.txt:1104]: Fair point.  Though this change shouldn't have happened in the first place.

File: builtin/mainmenu/dlg_create_world.lua
Comment (Cartmic) [reply builtin/mainmenu/dlg_create_world.lua:285]: Hi v-rob, it sound like a rebase, I don't recall changing this.

File: builtin/mainmenu/dlg_create_world.lua
Comment (Cartmic) [reply builtin/mainmenu/dlg_create_world.lua:76]: That's the first time I've seen a temples flag, so yes it sounds like a recent commit.

**Comment by Zughy** [comment]
Rebase mistakes weren't addressed. Closing for inactivity. If addressed, feel free to reopen

**Comment by ancientmarinerdev** [comment]
@v-rob I don't suppose you fancy adopting this? I appreciate you're very busy, but you understand it well and the changes needed. Know of anyone who would also be keen on this?

**Comment by v-rob** [comment]
Yes, I can pick this up. It should be simple enough for me to get a PR ready without too much trouble.

**Comment by mruncreative** [comment]
I don't like asking but what's the status of this?

**Pull request closed without merging**
